### ISSUE TITLE
Typedoc can't deal with one line in the new vega-lite package

### ISSUE THREAD

**Issue by jasongrout** [vega/vega-lite#3862]
Typedoc is currently constrained to use TypeScript 2.7 (upgrade issue at https://github.com/TypeStrong/typedoc/issues/733). We use typedoc to generate important API docs, and we also depend on vega-lite (through vega-embed). There is only one line that TS 2.7 chokes on in the vega-lite compiled package, which in turn is causing our api doc build to fail: 

```
Error: [snip]/node_modules/vega-lite/build/src/compositemark/index.d.ts(12)
 ',' expected.
```

That line is:
```
export declare const COMPOSITE_MARK_STYLES: import("./boxplot").BoxPlotStyle[];
```
Presumably it is generated from https://github.com/vega/vega-lite/blob/e2280723a89df327575c011784d6c7371216f1ec/src/compositemark/index.ts#L30

Do you think there might be any way to somehow get the vega-lite package to not generate TS 2.9-specific syntax, that `import("./boxplot").BoxPlotStyle[]`, in the .d.ts file? I realize it's a bit of a short-term hack (presumably would go away when typedoc is updated, which may be weeks or months away), but it lets people that depend on vega-lite to continue using typedoc, even with the newest version.

**Comment by domoritz** [comment]
I have never used typedoc and don't know exactly what it is chocking on. I'm happy to make temporary changes to Vega-Lite if necessary but I don't have the bandwidth to look into this issue soon. I happily review pull requests and make releases, though.

**Comment by jasongrout** [comment]
Thanks. I'll explore a bit to see if I can get Typescript to generate backwards-compatible code. I'm also seeing if we can move off of typedoc, if there is something for generating api documentation that keeps up to date better.

**Comment by domoritz** [comment]
In Vega-Lite, we use https://github.com/vega/ts-json-schema-generator to generate the schema and then we use Jekyll templates to generate the tables in the docs from the schema. This approach won't work for JupyterLab because you want API docs instead.

**Comment by jasongrout** [comment]
Thanks. I'm looking at some of the options at https://github.com/TypeStrong/typedoc/issues/701

**Comment by jasongrout** [comment]
This patch appears to fix the problem:

```patch
diff --git a/src/compositemark/index.ts b/src/compositemark/index.ts
index 6f089c70b..c80762d6b 100644
--- a/src/compositemark/index.ts
+++ b/src/compositemark/index.ts
@@ -4,6 +4,12 @@ import {GenericUnitSpec, NormalizedLayerSpec} from './../spec';
 import {BOXPLOT, BOXPLOT_STYLES, BoxPlotConfigMixins, BoxPlotDef, normalizeBoxPlot, VL_ONLY_BOXPLOT_CONFIG_PROPERTY_INDEX} from './boxplot';
 import {ERRORBAR, normalizeErrorBar} from './errorbar';
 
+// This package import below makes the generated .d.ts file compatible with
+// Typescript 2.7 so that libraries requiring us can use Typedoc (which
+// currently is limited to Typescript 2.7). This comment and import can be
+// removed when Typedoc is updated to Typescript 2.9 or later. See
+// https://github.com/vega/vega-lite/issues/3862 for more details.
+import * as boxplot from './boxplot';
 
 export {BoxPlotConfig} from './boxplot';
 export type UnitNormalizer = (spec: GenericUnitSpec<any, any>, config: Config)=> NormalizedLayerSpec;
```

**Comment by jasongrout** [comment]
In particular, with that patch, the offending line goes from:
```typescript
export declare const COMPOSITE_MARK_STYLES: import("./boxplot").BoxPlotStyle[];
```
to
```typescript
export declare const COMPOSITE_MARK_STYLES: boxplot.BoxPlotStyle[];
```

**Comment by jasongrout** [comment]
Thanks!

**Comment by jasongrout** [comment]
Fixed in #3863.
